acme/acme/challenges.py

```diff
@@ -27,16 +29,19 @@
 
 logger = logging.getLogger(__name__)
 
+GenericChallenge = TypeVar('GenericChallenge', bound='Challenge')
+
 
 class Challenge(jose.TypedJSONObjectWithFields):
     # _fields_to_partial_json
     """ACME challenge."""
     TYPES: Dict[str, Type['Challenge']] = {}
 
     @classmethod
-    def from_json(cls, jobj: Mapping[str, Any]) -> 'Challenge':
+    def from_json(cls: Type[GenericChallenge],
+                  jobj: Mapping[str, Any]) -> Union[GenericChallenge, 'UnrecognizedChallenge']:
         try:
-            return super().from_json(jobj)
+            return cast(GenericChallenge, super().from_json(jobj))
```

Comment:
We can do things this way, but do you think the need for a cast here suggest we may want to do something more/different with our generics in `josepy`?

Reply:
I thought about it a couple of time, and potentially we could also do things a little better. Here is the state of my analysis so far.

So, `TypeVar` is used for two usecases:
* being the type variable of a generic class (eg. `class MyList(Generic[T])` with `T`  a `TypeVar` instance)
* being the type variable for a generic method (typically `def add(one: A, two: A) -> A`)

We are talking here about the second case, which is to potentially take advantage of genericity for the `TypedJSONObjectWithFields.from_json` method and by extension the overriden method in `Challenge`.

In `JSONObjectWithFields.from_json` the genericity is possible, because the method "returns an instance of the class on which `from_json` is called", which is expressed by a `TypeVar` instance type on the `cls` parameter and its returned type (`def from_json(cls: Type[T], ...) -> T`), and this instructs mypy to dynamically infer the actual type of the method in sub-classes calls (eg. `from_json` effectively returns `SubJson` if `SubJson` is a child class of `JSONObjectWithFields`) when types are checked.

However `TypedJSONObjectWithFields.from_json` implementation and purpose slightly differs from `JSONObjectWithFields.from_json`. It is "returns an instance of a registered subclass of the current class given the `dict` provided by the method" (with internally a resolution done from the field `type` in the `dict`).

So potentially, we could declare a `TypeVar` bound to `TypedJSONObjectWithFields` in josepy, and use it to make `register`, `get_type_cls` and `from_json` generic, which means here accepting/returning any sub-type of the current class type.

I think it is representing the actual usecase of `TypedJSONObjectWithFields`, what do you think ?

Reply:
I responded to this at https://github.com/certbot/josepy/pull/122#issuecomment-976964453.

Reply:
At the end, this conversation induced some unrelated changes for josepy, and the `cast` will stay.